lightning/src/ln/channelmanager.rs

```diff
@@ -3321,6 +3377,100 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 			self.finish_force_close_channel(failure);
 		}
 	}
+
+	fn set_payment_hash_secret_map(&self, payment_hash: PaymentHash, payment_preimage: Option<PaymentPreimage>, min_value_msat: Option<u64>, invoice_expiry_delta_secs: u32, user_payment_id: u64) -> Result<PaymentSecret, APIError> {
+		assert!(invoice_expiry_delta_secs <= 60*60*24*365); // Sadly bitcoin timestamps are u32s, so panic before 2106
+
+		let payment_secret = PaymentSecret(self.keys_manager.get_secure_random_bytes());
```

Comment:
> Super likely we'll need to replace with a new keysinterface method for AMP/PTLC support where you want to enforce a cryptographic relation between the payment path identifier and the packet redeeming secret.

Hmm, I wasn't under the impression it would matter? The only relation that matters for the payment secret for invoice-initiated payments is that it came from the invoice to authenticate the sender. I kinda assumed PTLC stuff will only impact the payment hash-payment preimage relation, but maybe I'm missing something obvious?

> Shorter term, is this API keysend-proof

IIUC, keysend implies reading the payment preimage from the final hop onion TLV, which you can't do with either the current or the new API. We need to have a completely separate HTLC-processing path for HTLCs identifying themselves as keysend.